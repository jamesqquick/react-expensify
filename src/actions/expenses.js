import uuid from 'uuid';
import database from '../firebase/firebase';

export const addExpense = (expense = {}) => ({
    type: 'ADD_EXPENSE',
    expense
});

//returning a function thanks to THUNK
export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            description = '',
            note = '',
            amount = '',
            createdAt = ''
        } = expenseData;

        const expense = { description, note, amount, createdAt };
        return database
            .ref(`users/${uid}/expenses`)
            .push(expense)
            .then(ref => {
                console.log('Expense saved to firebase successfully', ref);
                dispatch(
                    addExpense({
                        id: ref.key,
                        ...expense
                    })
                );
            });
    };
};

export const deleteExpense = ({ id = -1 } = {}) => ({
    type: 'DELETE_EXPENSE',
    id
});

export const startDeleteExpense = (id = -1) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database
            .ref(`users/${uid}/expenses/${id}`)
            .remove()
            .then(() => {
                dispatch(deleteExpense({ id }));
            });
    };
};

export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

export const startEditExpense = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;

        return database
            .ref(`users/${uid}/expenses/${id}`)
            .update(updates)
            .then(() => {
                dispatch(editExpense(id, updates));
            });
    };
};

export const setExpenses = expenses => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;

        return database
            .ref(`users/${uid}/expenses`)
            .once('value')
            .then(snapshot => {
                const expenses = [];
                snapshot.forEach(childSnap => {
                    expenses.push({
                        ...childSnap.val(),
                        id: childSnap.key
                    });
                });
                dispatch(setExpenses(expenses));
            });
    };
};
