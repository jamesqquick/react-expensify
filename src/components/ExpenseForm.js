import 'react-dates/initialize';
import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            amount: props.expense
                ? (props.expense.amount / 100).toString()
                : '0',
            note: props.expense ? props.expense.note : '',
            createdAt: moment(),
            calendarFocused: false,
            error: ''
        };
    }

    onDescriptionChange = e => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };

    onNoteChange = e => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };

    onAmountChange = e => {
        const amount = e.target.value;
        //start with - caret
        // \d digit, * for any amount
        //optional stuff in parenthesis followed by question mark
        //ends with - dollar sign
        const regex = /^\d{1,}(\.\d{0,2})?$/;
        if (!amount || amount.match(regex)) {
            this.setState(() => ({ amount }));
        }
    };

    onDateChange = createdAt => {
        //dont allow to clear date
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };

    onDateFocusChanged = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };

    onFormSubmit = e => {
        e.preventDefault();
        //Check description and amount
        if (!this.state.description || !this.state.amount) {
            //Set error
            this.setState(() => ({
                error: 'Make sure to fill in description and amount.'
            }));
        } else {
            //Clear error
            this.setState(() => ({
                error: ''
            }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    };

    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onFormSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onDateFocusChanged}
                        numberOfMonths={1}
                        isOutsideRange={day => false}
                    />
                    <textarea
                        placeholder="Add a note for expense"
                        onChange={this.onNoteChange}
                    />
                    <button>Add Expense</button>
                </form>
            </div>
        );
    }
}

export default ExpenseForm;
