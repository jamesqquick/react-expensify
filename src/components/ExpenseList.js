import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

//Basic component
export const ExpenseList = props => (
	<div>
		<h1>Expense List</h1>
		{props.expenses.length === 0 && <p>No Expenses</p>}
		{props.expenses.map(expense => {
			return <ExpenseListItem key={expense.id} {...expense} />;
		})}
	</div>
);

//mapping of state to props
const mapStateToProps = state => ({
	expenses: selectExpenses(state.expenses, state.filters)
});

//exported component with props
export default connect(mapStateToProps)(ExpenseList);
