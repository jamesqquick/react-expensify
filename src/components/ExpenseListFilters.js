import React from 'react';
import { connect } from 'react-redux';
import {
    setTextFilter,
    sortByDate,
    sortByAmount,
    setStartDate,
    setEndDate
} from '../actions/filters';

import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    };

    onTextChange = e => {
        this.props.setTextFilter(e.target.value);
    };

    onSortChange = e => {
        if (e.target.value === 'date') {
            this.props.sortByDate();
        } else if (e.target.value === 'amount') {
            this.props.sortByAmount();
        }
    };

    render() {
        return (
            <div>
                <input
                    type="text"
                    id="textInput"
                    value={this.props.filters.text}
                    onChange={this.onTextChange}
                />
                <select
                    value={this.props.filters.filterBy}
                    onChange={this.onSortChange}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onDateRangeFocusChanged}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    showClearDates={true}
                    startDateId="startDate"
                    endDateId="endDate"
                />
            </div>
        );
    }
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onDateRangeFocusChanged = calendarFocused => {
        this.setState(() => ({ calendarFocused }));
    };
}

const mapStateToProps = state => ({ filters: state.filters });

const mapDispatchToProps = dispatch => {
    return {
        setTextFilter: text => dispatch(setTextFilter(text)),
        sortByDate: () => dispatch(sortByDate()),
        sortByAmount: () => dispatch(sortByAmount()),
        setStartDate: startDate => dispatch(setStartDate(startDate)),
        setEndDate: endDate => dispatch(setEndDate(endDate))
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
