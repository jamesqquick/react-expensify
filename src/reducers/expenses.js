const expenseReducerDefaultState = [];

export default (state = expenseReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [...state, action.expense];
        case 'DELETE_EXPENSE':
            return state.filter(({ id }) => id !== action.id);
        case 'EDIT_EXPENSE':
            return state.map(item => {
                return item.id === action.id
                    ? { ...item, ...action.updates }
                    : item;
            });
        case 'SET_EXPENSES':
            return action.expenses;
        default:
            return state;
    }
};
