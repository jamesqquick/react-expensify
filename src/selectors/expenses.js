import moment from 'moment';

export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses
        .filter(expense => {
            const createdAtMoment = moment(expense.createdAt);
            const afterStartDate = startDate
                ? startDate.isSameOrBefore(createdAtMoment, 'day')
                : true;
            const beforeEndDate = endDate
                ? endDate.isSameOrAfter(createdAtMoment, 'day')
                : true;
            const textMatch = expense.description
                .toLowerCase()
                .includes(text.toLowerCase());

            return afterStartDate && beforeEndDate && textMatch;
        })
        .sort((a, b) => {
            if (sortBy === 'date') {
                return a.createdAt < b.createdAt ? 1 : -1;
            } else if (sortBy === 'amount') {
                return a.amount > b.amount ? 1 : -1;
            }
        });
};
