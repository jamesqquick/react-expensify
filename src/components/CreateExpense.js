import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses';

export class CreateExpense extends React.Component {
    onSubmit = expense => {
        this.props.startAddExpense(expense);
        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                <p>Add</p>
                <ExpenseForm onSubmit={this.onSubmit} />
            </div>
        );
    }
}

const mapDispatchToProps = dispatch => {
    return {
        startAddExpense: expense => dispatch(startAddExpense(expense))
    };
};

export default connect(undefined, mapDispatchToProps)(CreateExpense);
