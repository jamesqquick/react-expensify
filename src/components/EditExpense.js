import React from 'react';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startDeleteExpense } from '../actions/expenses';
import { connect } from 'react-redux';

export class EditExpense extends React.Component {
    render() {
        return (
            <div>
                Edit expense: {this.props.match.params.id}
                <ExpenseForm
                    onSubmit={this.submitEdit}
                    expense={this.props.expense}
                />
                <button
                    id="startDeleteExpenseBtn"
                    onClick={this.startDeleteExpense}
                >
                    Delete
                </button>
            </div>
        );
    }

    submitEdit = expense => {
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    };

    startDeleteExpense = () => {
        this.props.startDeleteExpense(this.props.expense.id);
        this.props.history.push('/');
    };
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find(
            expense => expense.id === props.match.params.id
        )
    };
};

const mapDispatchToProps = dispatch => {
    return {
        startEditExpense: (id, expense) =>
            dispatch(startEditExpense(id, expense)),
        startDeleteExpense: data => dispatch(startDeleteExpense(data))
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpense);
