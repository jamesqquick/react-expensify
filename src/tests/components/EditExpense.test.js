import React from 'react';
import { shallow } from 'enzyme';
import { EditExpense } from '../../components/EditExpense';
import expenses from '../fixtures/expenses';

let startEditExpense, startDeleteExpense, history, wrapper, match;

beforeEach(() => {
    startEditExpense = jest.fn();
    startDeleteExpense = jest.fn();
    history = { push: jest.fn() };
    match = {
        params: {
            id: expenses[0].id
        }
    };

    wrapper = shallow(
        <EditExpense
            startEditExpense={startEditExpense}
            startDeleteExpense={startDeleteExpense}
            expense={expenses[0]}
            history={history}
            match={match}
        />
    );
});

test('should render create expense page properly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should handle on submit', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startEditExpense).toHaveBeenLastCalledWith(
        expenses[0].id,
        expenses[0]
    );
});

test('should handle delete expense', () => {
    wrapper.find('#startDeleteExpenseBtn').simulate('click');
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startDeleteExpense).toHaveBeenCalledWith(expenses[0].id);
});
