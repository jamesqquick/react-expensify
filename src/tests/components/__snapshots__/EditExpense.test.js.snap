// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render create expense page properly 1`] = `
<div>
  Edit expense: 
  1
  <ExpenseForm
    expense={
      Object {
        "amount": 195,
        "createdAt": 0,
        "description": "First Note",
        "id": "1",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    id="startDeleteExpenseBtn"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
